File: OPL/JAVA/oplcanoga/app/src/main/java/com/example/oplcanoga/model/ComputerPlayer.java
package com.example.oplcanoga.model;

import java.util.Comparator;
import java.util.List;

/**
 * Represents the computer (AI) player.
 * Contains logic for making decisions about rolling dice and choosing moves.
 */
public class ComputerPlayer extends Player {

    // Helper field to store the reason for the last decision
    private String lastMoveReason = "";

    /**
     * Constructs a ComputerPlayer with the specified board size.
     *
     * @param boardSize The number of squares on the player's board.
     */
    public ComputerPlayer(int boardSize) {
        super(PlayerId.COMPUTER, boardSize);
    }

    /**
     * Decides whether to roll one die or two.
     * The strategy is to roll one die whenever allowed (which generally maximizes the chance of covering small numbers needed at the end).
     *
     * @param round The current game round context.
     * @return True to roll one die, false to roll two.
     */
    public boolean decideRollOneDie(GameRound round) {
        return round.canRollOneDie(PlayerId.COMPUTER);
    }

    /**
     * Retrieves the explanation for the last chosen move.
     *
     * @return A string explaining the strategy behind the last move.
     */
    public String getLastMoveReason() {
        return lastMoveReason;
    }

    /**
     * Chooses the best move from the available legal moves.
     * <p>
     * Strategy:
     * 1. If there's a winning move (covers all remaining squares or uncovers all opponent squares), take it.
     * 2. Otherwise, prefer covering moves over uncovering moves.
     * 3. Among the chosen type, maximize the number of squares involved, then the highest square value.
     * </p>
     *
     * @param coverMoves   List of legal moves that cover squares.
     * @param uncoverMoves List of legal moves that uncover squares.
     * @param me           The computer player.
     * @param opponent     The opponent player.
     * @return The chosen Move object, or null if no moves are available.
     */
    public Move chooseMove(List<Move> coverMoves, List<Move> uncoverMoves, Player me, Player opponent) {
        lastMoveReason = ""; // Reset reason

        // 1. Check for immediate win by covering
        int myUncoveredCount = 0;
        for (int i = 1; i <= me.getBoardSize(); i++) {
            if (!me.isCovered(i)) myUncoveredCount++;
        }
        for (Move m : coverMoves) {
            if (m.getSquareCount() == myUncoveredCount) {
                lastMoveReason = "I can win immediately by covering all my remaining squares!";
                return m;
            }
        }

        // 2. Check for immediate win by uncovering
        int oppCoveredCount = 0;
        for (int i = 1; i <= opponent.getBoardSize(); i++) {
            if (opponent.isCovered(i)) oppCoveredCount++;
        }
        for (Move m : uncoverMoves) {
            if (m.getSquareCount() == oppCoveredCount) {
                lastMoveReason = "I can win immediately by uncovering all of your covered squares!";
                return m;
            }
        }

        // 3. Default strategy: Prefer covering, then maximizing count/value
        List<Move> candidates;
        boolean preferCover = !coverMoves.isEmpty();
        
        if (preferCover) {
            candidates = coverMoves;
        } else {
            candidates = uncoverMoves;
        }

        if (candidates.isEmpty()) {
            lastMoveReason = "No legal moves available.";
            return null;
        }

        Move bestMove = candidates.stream()
                .max(Comparator
                        .comparingInt(Move::getSquareCount)
                        .thenComparingInt(Move::getHighestSquare))
                .orElse(null);
        
        if (bestMove != null) {
            if (preferCover) {
                lastMoveReason = "Strategy: Covering my squares (" + bestMove.getSquares() + ") reduces your potential score. " +
                        "This option covers the most squares or highest values among available moves.";
            } else {
                lastMoveReason = "Strategy: Uncovering your squares (" + bestMove.getSquares() + ") makes it harder for you to win. " +
                        "This option affects the most squares or highest values.";
            }
        }
        
        return bestMove;
    }
}

File: OPL/JS_FINAL/js/model/ComputerPlayer.js
// js/model/ComputerPlayer.js
import { Player } from "./Player.js";

/**
 * ComputerPlayer encapsulates AI strategy and help-mode logic.
 *
 * It does NOT roll dice or control turn flow.
 * It only decides:
 *  - cover vs uncover
 *  - which squares to use
 */
export class ComputerPlayer extends Player {
  constructor(name = "Computer") {
    super("COMPUTER", name);
  }

  /**
   * Decide best move for given player (usually the computer itself).
   * We keep this generic so it can also be used to give help to human.
   *
   * SMART STRATEGY:
   *  1. Check ALL legal moves (cover + uncover) for an IMMEDIATE WIN.
   *     - Cover: move covers ALL your remaining uncovered squares.
   *     - Uncover: move uncovers ALL of opponent's remaining covered squares.
   *  2. If no winning move:
   *     - Prefer covering over uncovering (rubric rule).
   *     - Within cover/uncover moves, choose combo with:
   *         a) more squares,
   *         b) if tie, larger sum.
   *
   * @param {GameRound} round
   * @param {string} playerId "HUMAN" | "COMPUTER"
   * @param {number} diceSum
   * @returns {{
   *   action: "cover" | "uncover" | "none",
   *   squares: number[],
   *   reason: string,
   *   allCoverOptions: number[][],
   *   allUncoverOptions: number[][]
   * }}
   */
  decideMove(round, playerId, diceSum) {
    const coverOptions = round.getCoverOptions(playerId, diceSum);
    const uncoverOptions = round.getUncoverOptions(playerId, diceSum);

    if (coverOptions.length === 0 && uncoverOptions.length === 0) {
      return {
        action: "none",
        squares: [],
        reason: "No valid cover or uncover moves available for this dice sum.",
        allCoverOptions: [],
        allUncoverOptions: []
      };
    }

    const self = round.getPlayerById(playerId);
    const opp = round.getOpponent(playerId);

    const selfUncovered = self.board.getUncoveredNumbers();
    const oppCovered = opp.board.getCoveredNumbers();

    // ------------------------------------------------------
    // 1. LOOK FOR AN IMMEDIATE WIN (COVER OR UNCOVER)
    // ------------------------------------------------------

    // Win by COVER: if this combo covers ALL remaining uncovered squares
    // => combo length == #uncovered (combos are subsets of uncovered).
    for (const combo of coverOptions) {
      if (combo.length === selfUncovered.length) {
        return {
          action: "cover",
          squares: combo,
          reason:
            "This move COVERS all your remaining squares and wins the round immediately.",
          allCoverOptions: coverOptions,
          allUncoverOptions: uncoverOptions
        };
      }
    }

    // Win by UNCOVER: if this combo uncovers ALL of opponent's covered squares
    // => combo length == #covered (combos are subsets of covered).
    for (const combo of uncoverOptions) {
      if (combo.length === oppCovered.length) {
        return {
          action: "uncover",
          squares: combo,
          reason:
            "This move UNCOVERS all opponent squares and wins the round immediately.",
          allCoverOptions: coverOptions,
          allUncoverOptions: uncoverOptions
        };
      }
    }

    // ------------------------------------------------------
    // 2. NO FORCED WIN: USE HEURISTIC
    //    - Prefer covering if possible.
    //    - Otherwise uncover.
    //    - Within each, prefer:
    //        * more squares
    //        * if tie, larger sum
    // ------------------------------------------------------
    if (coverOptions.length > 0) {
      const best = this._pickBestCombo(coverOptions);
      return {
        action: "cover",
        squares: best,
        reason:
          "No immediate win available. Chose to COVER own squares using the combination that makes the most progress (more / larger squares).",
        allCoverOptions: coverOptions,
        allUncoverOptions: uncoverOptions
      };
    } else {
      const best = this._pickBestCombo(uncoverOptions);
      return {
        action: "uncover",
        squares: best,
        reason:
          "No immediate win available and no cover move. Chose to UNCOVER opponent's squares using the strongest combination (more / larger squares).",
        allCoverOptions: coverOptions,
        allUncoverOptions: uncoverOptions
      };
    }
  }

  /**
   * Helper for help-mode: same logic as decideMove, but explicitly labelled.
   */
  getHelpSuggestion(round, playerId, diceSum) {
    // Help is literally "what would the AI do in your shoes?"
    return this.decideMove(round, playerId, diceSum);
  }

  /**
   * Choose whether to roll 1 die or 2 dice based on board state.
   *
   * Rule (per rubric):
   *  - If 7..n are all covered: roll 1 die.
   *  - Otherwise: roll 2 dice.
   *
   * @param {Board} board the player's own board
   * @returns {1 | 2}
   */
  chooseNumDice(board) {
    const n = board.size;
    let allHighCovered = true;
    for (let i = 7; i <= n; i++) {
      if (i >= 1 && i <= n && !board.isCovered(i)) {
        allHighCovered = false;
        break;
      }
    }
    return allHighCovered ? 1 : 2;
  }

  /**
   * Generic "best combo" according to:
   *  - more squares is better
   *  - if tie, higher total sum is better
   */
  _pickBestCombo(options) {
    let best = options[0];
    for (let i = 1; i < options.length; i++) {
      const candidate = options[i];
      if (candidate.length > best.length) {
        best = candidate;
      } else if (candidate.length === best.length) {
        const sumBest = best.reduce((a, b) => a + b, 0);
        const sumCand = candidate.reduce((a, b) => a + b, 0);
        if (sumCand > sumBest) {
          best = candidate;
        }
      }
    }
    return best;
  }
}
